CHANGELOG.md

```diff
@@ -33,6 +34,19 @@
 - [v0.2.0](#v020)
 - [v0.1.0](#v010)
 
+## [v1.8.3]
+
+> Release date: 2021/12/02
+
+### Fixed
+
+- `--select-tag` applies its tags to newly-created entities whether or not the
+  tag is also present in the state file metadata.
+  [#517](https://github.com/Kong/deck/pull/517)
```

Comment:
Isn't this a breaking change that would command more meaningful than a patch version bump? :thinking:

Reply:
"Isn't this a breaking change?" absent analysis of the potential breakage isn't of much use. The Hyrum article's advice is "be aware of possible implicit interfaces", not "every change in behavior is guaranteed breaking, you can never make patch releases ever". The latter reductio ad absurdum approach means we should declare any code change breaking. Fix a segfault? It's a breaking change, someone relied on that segfault occurring.

I don't think we have any good hypotheticals for this, so IMO labeling it breaking is more confusing than anything: messaging to users is that it's a breaking change because it's a change, but we aren't aware of any situations where it'd break anything, so we can't offer any advice on what you'd need to adjust beyond "run tests and see if they fail", which they should be doing anyway. Based on what we do know, this change does not break expected workflows, and we should mark it as such, i.e. our messaging should be to users that we do not expect any negative impact or changes required if you're coming from another 1.8.x release.

A workflow that relied on the existing behavior would need both an intimate understanding of the behavior and an intentionally contrived workflow to rely on it. The current behavior de facto makes `--select-tag` without file metadata useless for creating new entities. If you run it twice, the created entities are invisible and deck tries to create them unsuccessfully:

```
$ /tmp/bdeck sync -s /tmp/xxx.yaml --select-tag=foo
creating upstream awesome_learnservice
...

$ /tmp/bdeck sync -s /tmp/xxx.yaml --select-tag=foo
creating upstream awesome_learnservice
Summary:
  Created: 0
  Updated: 0
  Deleted: 0
Error: 1 errors occurred:
	while processing event: {Create} upstream awesome_learnservice failed: HTTP status 409 (message: "UNIQUE violation detected on '{name=\"awesome_learnservice\"}'")
```
Expected behavior (repeated syncs of the same content succeed and do not modify resources after the first sync) only works with `--select-tag` and no file metadata if tags are already present on all resources. This is unaffected by the change, since those resources already had tags listed in their individual YAML objects: 
[existing_tag.txt](https://github.com/Kong/deck/files/7644282/existing_tag.txt) (bdeck is 1.8.2, ndeck is this branch).

As best I can reason, to rely on the existing behavior, you'd need to intentionally either initially dump without `--select-tag` (so you get existing resources but no file tag metadata) or write a state file with no tag metadata from scratch, append new objects, sync with `--select-tag`, know that the newly-created objects would have no tags, and then run subsequent syncs without `--select-tag` to update those untagged objects. At that point, it's unclear why you wouldn't have run without `--select-tag` in the first place, because you'd wind up with the same result. Even if you did that, the breakage is that now your objects do have tags, which makes no difference--the sync you'd have run after, with no `--select-tag`, will manipulate them regardless, because syncs without tag criteria match any tag set.

The only (bizarre, and based on practical testing, also impossible) workflow I can think of where you could utilize the existing behavior and would be broken with this change is if you wanted to create a new object unconditionally if no matching object with the matching tag exists. Effectively:

1. You create a state file without tag metadata and some object `foo`. `foo` must not contain any unique fields (and by extension, its type's schema cannot have any required unique fields).
2. You run `deck sync --select-tag whatever`. If an object matching `foo` exists with tag `whatever`, deck updates that object with any changes needed, or does nothing if no changes are needed. If no object matching foo exists with tag `whatever`, deck creates a new object without the tag `whatever`.
3. Assuming the second case (new untagged object), subsequent runs with the same state and parameters will create additional `foo` objects, one for each run.

There are some types that match this schema constraint, e.g. you can create multiple routes with nothing other than `hostnames=[example.com]`. Routes don't require names or other unique fields (other than ID, which deck will generate automatically). However, you can't actually do this with deck: deck imposes additional constraints beyond the Kong schema:

```
$ /tmp/bdeck sync -s /tmp/wat.yaml --select-tag foo
Error: reading file: validating file content: 2 errors occurred:
	routes.0: Must validate at least one schema (anyOf)
	routes.0: name is required
```
deck does require at least one unique field on such objects intentionally to ensure that the entire Kong config in scope matches state; it's intentionally limited so that syncs are idempotent and cannot create multiple copies of a resource in state.

Reply:
> I don't think we have any good hypotheticals for this, so IMO labeling it breaking is more confusing than anything

SGTM.

Reply:
> The latter reductio ad absurdum approach means we should declare any code change breaking. Fix a segfault? It's a breaking change, someone relied on that segfault occurring.

I disagree. Changes that do modify the public interface are breaking. A segfault is not a part of the public interface of deck; it's an unrecoverable runtime error.

> bizarre, and based on practical testing, also impossible

This is convincing. I wasn't sure about the impact of this change on decK's semantics, hence my question - I initially understood a scenario where someone would suffer from a tags mismatch handling difference to be more likely.